lib/mayaUsd/utils/converter.h

```diff
@@ -0,0 +1,331 @@
+    void convert(const UsdAttribute& src, MPlug& plug, MDGModifier& dst, const ConverterArgs& args) const
+    { _attrToModifier(src,plug,dst,args); }
+    
+    void convert(const MPlug& src, VtValue& dst, const ConverterArgs& args) const
+    { _plugToVtValue(src, dst, args); }
+    void convert(const VtValue& src, MPlug& dst, const ConverterArgs& args) const
+    { _vtValueToPlug(src, dst, args); }
+    void convert(const VtValue& src, MPlug& plug, MDGModifier& dst, const ConverterArgs& args) const
+    { _vtValueToModifier(src, plug, dst, args); }
+    
+    void convert(const MDataHandle& src, UsdAttribute& dst, const ConverterArgs& args) const
+    { _handleToAttr(src,dst,args); }
+    void convert(const UsdAttribute& src, MDataHandle& dst, const ConverterArgs& args) const
+    { _attrToHandle(src,dst,args); }
+    
+    void convert(const MDataHandle& src, VtValue& dst, const ConverterArgs& args) const
+    { _handleToVtValue(src, dst, args); }
+    void convert(const VtValue& src, MDataHandle& dst, const ConverterArgs& args) const
+    { _vtValueoHandle(src, dst, args); }
+    
+    static bool hasAttrType(const MPlug& plug, MFnData::Type type)
+    {
+        MObject object = plug.attribute();
+        if (!object.hasFn(MFn::kTypedAttribute)) {
+            return false;
+        }
+
+        MFnTypedAttribute attr(object);
+        return attr.attrType() == type;
+    }
+
+    static bool hasNumericType(const MPlug& plug, MFnNumericData::Type type)
+    {
+        MObject object = plug.attribute();
+        if (!object.hasFn(MFn::kNumericAttribute)) {
+            return false;
+        }
+
+        MFnNumericAttribute attr(object);
+        return attr.unitType() == type;
+    }
+    
+    static SdfValueTypeName getUsdTypeName(
+        const MPlug& attrPlug,
+        const bool translateMayaDoubleToUsdSinglePrecision);
+
+private:
+    const SdfValueTypeName& _typeName;
+    
+    // MPlug <--> UsdAttribute
+    MPlugToUsdAttrFn _plugToAttr{nullptr};
+    UsdAttrToMPlugFn _attrToPlug{nullptr};
+    UsdAttrToMDGModifierFn _attrToModifier{nullptr};
+    // MPlug <--> VtValue
+    MPlugToVtValueFn _plugToVtValue{nullptr};
+    VtValueToMPlugFn _vtValueToPlug{nullptr};
+    VtValueToMDGModifierFn _vtValueToModifier{nullptr};
+    
+    // MDataBlock <--> UsdAttribute
+    MDataHandleToUsdAttrFn _handleToAttr{nullptr};
+    UsdAttrToMDataHandleFn _attrToHandle{nullptr};
+    // MDataBlock <--> VtValue
+    MDataHandleToVtValueFn _handleToVtValue{nullptr};
+    VtValueToMDataHandleFn _vtValueoHandle{nullptr};
+};
+
+template <class MAYA_Type, class USD_Type, class Enable=void> struct TypedConverter {};
+
+template <class MAYA_Type, class USD_Type>
+struct TypedConverter<MAYA_Type,USD_Type, typename std::enable_if<std::is_same<MAYA_Type, USD_Type>::value>::type>
+{
+    static void convert(const USD_Type& src, MAYA_Type& dst)
+    {
+        dst = src;
+    }
+};
+
+template <>
+struct TypedConverter<float3,GfVec3f>
```

Comment:
Curious to hear your thoughts about how types that can be safely converted implicitly might look here.

For example:
```
src = float3, dst = GfVec3d    # OK
src = double3, dst = GfVec3f   # INVALID, potential data loss
```

Reply:
I hope this comment shares my current thinking https://github.com/Autodesk/maya-usd/pull/432#discussion_r425930844